File: senior/MT1051/week05/HW12/HW12_109601003.go
/*
Created Date: 2023/10/16
HW: 12
Author: 林群賀
Student Number: 109601003
*/

package main

import (
	"fmt"
)

func targetFunction1(x float64) float64 {
	return 3 * x * x - 6 * x + 3
}

func findMin1(startX, endX, step float64) (float64, float64) {
	minX := startX
	minY := targetFunction1(startX)

	currentX := startX

	for currentX <= endX {
		currentY := targetFunction1(currentX)
		if currentY < minY {
			minX = currentX
			minY = currentY
		}
		currentX += step
	}

	return minX, minY
}

func targetFunction2(x float64) float64 {
	return 0.5*x*x - 3*x + 7.5
}

func findMin2(startX, endX, step float64) (float64, float64) {
	minX := startX
	minY := targetFunction2(startX)

	currentX := startX

	for currentX <= endX {
		currentY := targetFunction2(currentX)
		if currentY < minY {
			minX = currentX
			minY = currentY
		}
		currentX += step
	}

	return minX, minY
}

func main() {
	startX := -10.0
	endX := 10.0
	step := 0.01

	minX1, minY1 := findMin1(startX, endX, step)
	minX2, minY2 := findMin2(startX, endX, step)

	fmt.Printf("(1) min: x = %f, y = %f\n", minX1, minY1)
	fmt.Printf("(2) min: x = %f, y = %f\n", minX2, minY2)
}

File: mt/mt1051/week05/HW12/HW12_109601003.py
# -*- coding: utf-8 -*-
"""
Date: 2023/10/16
HW: 12
Author: 林群賀
Student Number: 109601003
"""

# # Define the target function
# def target_function(x):
#     return 3 * x * x - 6 * x + 3

# # Define the range over which you want to search for the minimum
# # Adjust the range and step size as needed
# start_x = -10
# end_x = 10
# step = 0.01

# min_x = start_x
# min_y = target_function(start_x)

# for x in range(int(start_x * 100), int(end_x * 100)):
#     current_x = x / 100
#     current_y = target_function(current_x)
#     if current_y < min_y:
#         min_x = current_x
#         min_y = current_y

# print(f"for loop - min: x = {min_x}, y = {min_y}")

# Define the target function
def target_function(x):
    return 3 * x * x - 6 * x + 3

start_x = -10
end_x = 10
step = 0.01

min_x = start_x
min_y = target_function(start_x)

current_x = start_x

while current_x <= end_x:
    current_y = target_function(current_x)
    if current_y < min_y:
        min_x = current_x
        min_y = current_y
    current_x += step

print(f"(1). min: x = {min_x}, y = {min_y}")

# Define the target function
def target_function(x):
    return 1/2 * x * x - 3 * x + 7/2

# Define the range over which you want to search for the minimum
# Adjust the range and step size as needed
start_x = -10
end_x = 10
step = 0.01

min_x = start_x
min_y = target_function(start_x)

current_x = start_x

while current_x <= end_x:
    current_y = target_function(current_x)
    if current_y < min_y:
        min_x = current_x
        min_y = current_y
    current_x += step

print(f"(2). min: x = {min_x}, y = {min_y}")
